### PR TITLE
Fix jitlayers aotcompile

### PR DESCRIPTION
**Pull request by qinsoon**
This PR fixes a few issues in `jitlayers.cpp` and `aotcompile.cpp` so they are compliant with the GC checker. More logging and tests for GC checker.

### CODE REVIEW

File: src/aotcompile.cpp [hunk 1]
```diff
@@ -240,8 +240,13 @@ static void jl_ci_cache_lookup(const jl_cgparams_t &cgparams, jl_method_instance
         jl_method_t *def = codeinst->def->def.method;
         if ((jl_value_t*)*src_out == jl_nothing)
             *src_out = NULL;
-        if (*src_out && jl_is_method(def))
+        if (*src_out && jl_is_method(def)) {
+            PTR_PIN(def);
+            PTR_PIN(codeinst);
             *src_out = jl_uncompress_ir(def, codeinst, (jl_array_t*)*src_out);
```
Comment (udesou) [inline src/aotcompile.cpp:246]: Why we don't need to pin `*src_out`, as it seems it's been cast to `jl_array_t*`?

Comment (qinsoon) [reply src/aotcompile.cpp:246]: I am looking at this. It looks like a bug. The arg `jl_value_t **src_out` is pinned, but we cannot say `*src` is pinned.

**Comment by qinsoon** [comment]
@udesou 

I fixed some issues in GCChecker. Can you take a look at the PR again? You can focus on the changes about the test cases, regarding to pinning, which are simple cases, and the changes in `jitlayers.cpp` and `aotcompile.cpp`. There could still be more issues with GCChecker -- if you find any, just let me know.

I am not sure if we want to merge this PR at all. Maybe we should only merge changes in GCChecker so we can use it to check the runtime code. We keep runtime changes in a separate branch. We only merge runtime changes after we go through the entire runtime code base, fixing pinning issues and verify correctness and performance. What do you think?

**Comment by qinsoon** [comment]
> @udesou
> 
> I fixed some issues in GCChecker. Can you take a look at the PR again? You can focus on the changes about the test cases, regarding to pinning, which are simple cases, and the changes in `jitlayers.cpp` and `aotcompile.cpp`. There could still be more issues with GCChecker -- if you find any, just let me know.
> 
> I am not sure if we want to merge this PR at all. Maybe we should only merge changes in GCChecker so we can use it to check the runtime code. We keep runtime changes in a separate branch. We only merge runtime changes after we go through the entire runtime code base, fixing pinning issues and verify correctness and performance. What do you think?

I created two separate PRs. https://github.com/mmtk/julia/pull/51 fixes the GCChecker, which I suggest to be merged as soon as we can. And https://github.com/mmtk/julia/pull/52 fixes the runtime code, which I suggest to remain a draft until we go through and fix the entire code base.

I am closing this PR.

**Pull request closed without merging**